Pass Vault agent annotations to Consul API Gateway controller

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

<!--- Thank you for keeping this note for the community --->

---

<!--- When filing a bug, please include the following headings if possible. Any example text in this template can be deleted. --->

### Overview of the Issue

<!--- Please describe the issue you are having and how you encountered the problem. --->

I deployed the Consul Helm chart with the Vault secrets backend and Consul API Gateway enabled. The Consul API Gateway was stuck initializing because the Vault agent needed to authenticate to HCP Vault, which requires a Vault namespace annotation.

`.Values.global.secretsBackend.vault.agentAnnotations` must to be added to [api-gateway-controller-deployment.yaml](https://github.com/hashicorp/consul-k8s/blob/v0.44.0/charts/consul/templates/api-gateway-controller-deployment.yaml#L26) to allow the Vault agent to properly authenticate to Vault.

### Reproduction Steps

<!--- 

In order to effectively and quickly resolve the issue, please provide exact steps that allow us the reproduce the problem. If no steps are provided, then it will likely take longer to get the issue resolved. An example that you can follow is provided below. 

Steps to reproduce this issue, eg:

1. When running helm install with the following `values.yml`:
```
global:
  domain: consul
  datacenter: dc1
server:
  replicas: 1
  bootstrapExpect: 1
connectInject:
  enabled: true
controller:
  enabled: true
```
  --->

1. Run `helm install` with the following `values.yml`:
  
  ```
  global:
    datacenter: "${CONSUL_DATACENTER}"
    name: consul
    secretsBackend:
      vault:
        enabled: true
        consulServerRole: ${CONSUL_SERVER_ROLE}
        consulClientRole: ${CONSUL_CLIENT_ROLE}
        consulCARole: ${CONSUL_CA_ROLE}
        manageSystemACLsRole: ${SERVER_ACL_INIT_ROLE}
        agentAnnotations: |
          "vault.hashicorp.com/namespace": "${VAULT_NAMESPACE}"
        connectCA:
          address: ${VAULT_ADDR}
          rootPKIPath: ${CONSUL_CONNECT_PKI_PATH_ROOT}
          intermediatePKIPath: ${CONSUL_CONNECT_PKI_PATH_INT}
          authMethodPath: ${KUBERNETES_AUTH_METHOD_PATH}
          additionalConfig: '"{"connect": [{ "ca_config": [{ "namespace": "${VAULT_NAMESPACE}"}]}]}"'
    tls:
      enabled: true
      enableAutoEncrypt: true
      caCert:
        secretName: "${CONSUL_PKI_PATH}/cert/ca"
      caKey:
        secretName: "${CONSUL_PKI_PATH}/issue/${CONSUL_SERVER_ROLE}"
        secretKey: private_key
    acls:
      manageSystemACLs: true
      bootstrapToken:
        secretName: "${CONSUL_STATIC_PATH}/data/bootstrap"
        secretKey: token
    gossipEncryption:
      secretName: ${CONSUL_STATIC_PATH}/data/gossip
      secretKey: key
  
  server:
    replicas: 1
    serverCert:
      secretName: "${CONSUL_PKI_PATH}/issue/${CONSUL_SERVER_ROLE}"
  
  apiGateway:
    enabled: true
    image: "hashicorp/consul-api-gateway:0.2.1"
    managedGatewayClass:
      serviceType: LoadBalancer
      useHostPorts: true
  ```

3. The Consul API Gateway controller is stuck initializing.

```shell
$ kubectl get pods

NAME                                             READY   STATUS     RESTARTS   AGE
consul-api-gateway-controller-849649d5b8-vns64   0/2     Init:0/4   0          48s
```

### Logs

<!---

Provide log files from Consul Kubernetes components by providing output from `kubectl logs` from the pod and container that is surfacing the issue. 

--->

<details>
  <summary>Logs</summary>

```
$ kubectl logs consul-api-gateway-controller-849649d5b8-vns64 -c vault-agent-init
==> Vault agent started! Log data will stream in below:

==> Vault agent configuration:

                     Cgo: disabled
               Log Level: info
                 Version: Vault v1.10.3
             Version Sha: af866591ee60485f05d6e32dd63dde93df686dfb

2022-05-26T20:19:00.146Z [INFO]  sink.file: creating file sink
2022-05-26T20:19:00.146Z [INFO]  sink.file: file sink configured: path=/home/vault/.vault-token mode=-rw-r-----
2022-05-26T20:19:00.146Z [INFO]  auth.handler: starting auth handler
2022-05-26T20:19:00.146Z [INFO]  auth.handler: authenticating
2022-05-26T20:19:00.147Z [INFO]  template.server: starting template server
2022-05-26T20:19:00.147Z [INFO]  sink.server: starting sink server
2022-05-26T20:19:00.147Z [INFO] (runner) creating new runner (dry: false, once: false)
2022-05-26T20:19:00.148Z [INFO] (runner) creating watcher
2022-05-26T20:19:00.164Z [ERROR] auth.handler: error authenticating:
  error=
  | Error making API request.
  | 
  | URL: PUT https://REDACTED:8200/v1/auth/kubernetes/login
  | Code: 403. Errors:
  | 
  | * permission denied
   backoff=1s
```

</details>

### Expected behavior

<!--- What was the expected result after following the reproduction steps? --->

Consul API Gateway should be running.

### Environment details

<!---

If not already included, please provide the following:
- `consul-k8s` version:
- `values.yaml` used to deploy the helm chart:

Additionally, please provide details regarding the Kubernetes Infrastructure, as shown below:
- Kubernetes version: v1.22.x
- Cloud Provider (If self-hosted, the Kubernetes provider utilized): EKS, AKS, GKE, OpenShift (and version), Rancher (and version), TKGI (and version)
- Networking CNI plugin in use: Calico, Cilium, NSX-T 

Any other information you can provide about the environment/deployment.
--->

- Consul Helm chart 0.44.0
- Consul 1.12.0
- HCP Vault (Vault version 1.10.3)
- Kubernetes v1.22.6-eks-14c7a48
- Cloud Provider EKS

### Additional Context

<!---
Additional context on the problem. Docs, links to blogs, or other material that lead you to discover this issue or were helpful in troubleshooting the issue. 
--->

Currently, I am patching the deployment with:

```shell
kubectl patch deployment consul-api-gateway-controller -p '{"spec": {"template":{"metadata":{"annotations":{"vault.hashicorp.com/namespace":"admin"}}}}}'
```

This allows the Vault agent container to authenticate correctly to Vault and inject the certificates for the controller.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,7 @@ IMPROVEMENTS:
 BUG FIXES:
 * Helm
   * Update client-snapshot-agent so that setting `client.snapshotAgent.caCert` no longer requires root access to modify the trust store. [[GH-1190](https://github.com/hashicorp/consul-k8s/pull/1190/)]
+  * Add missing vault agent annotations to the `api-gateway-controller-deployment`. [[GH-1247](https://github.com/hashicorp/consul-k8s/pull/1247)]
 
 ## 0.44.0 (May 17, 2022)
 

--- a/charts/consul/templates/api-gateway-controller-deployment.yaml
+++ b/charts/consul/templates/api-gateway-controller-deployment.yaml
@@ -31,6 +31,9 @@ spec:
         "vault.hashicorp.com/role": {{ .Values.global.secretsBackend.vault.consulCARole }}
         "vault.hashicorp.com/agent-inject-secret-serverca.crt": {{ .Values.global.tls.caCert.secretName }}
         "vault.hashicorp.com/agent-inject-template-serverca.crt": {{ template "consul.serverTLSCATemplate" . }}
+        {{- if .Values.global.secretsBackend.vault.agentAnnotations }}
+        {{ tpl .Values.global.secretsBackend.vault.agentAnnotations . | nindent 8 | trim }}
+        {{- end }}
         {{- if and .Values.global.secretsBackend.vault.ca.secretName .Values.global.secretsBackend.vault.ca.secretKey }}
         "vault.hashicorp.com/agent-extra-secret": "{{ .Values.global.secretsBackend.vault.ca.secretName }}"
         "vault.hashicorp.com/ca-cert": "/vault/custom/{{ .Values.global.secretsBackend.vault.ca.secretKey }}"

--- a/charts/consul/test/unit/api-gateway-controller-deployment.bats
+++ b/charts/consul/test/unit/api-gateway-controller-deployment.bats
@@ -829,3 +829,22 @@ load _helpers
       yq -r '.annotations["vault.hashicorp.com/role"]' | tee /dev/stderr)"
   [ "${actual}" = "test" ]
 }
+
+@test "apiGateway/Deployment: vault agent annotations can be set" {
+  cd `chart_dir`
+  local actual=$(helm template \
+      -s templates/api-gateway-controller-deployment.yaml  \
+      --set 'apiGateway.enabled=true' \
+      --set 'apiGateway.image=foo' \
+      --set 'global.tls.enabled=true' \
+      --set 'global.tls.enableAutoEncrypt=true' \
+      --set 'global.tls.caCert.secretName=foo' \
+      --set 'global.secretsBackend.vault.enabled=true' \
+      --set 'global.secretsBackend.vault.consulClientRole=test' \
+      --set 'global.secretsBackend.vault.consulServerRole=foo' \
+      --set 'global.secretsBackend.vault.consulCARole=test' \
+      --set 'global.secretsBackend.vault.agentAnnotations=foo: bar' \
+      . | tee /dev/stderr |
+      yq -r '.spec.template.metadata.annotations.foo' | tee /dev/stderr)
+  [ "${actual}" = "bar" ]
+}

